src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py

```diff
@@ -0,0 +1,357 @@
+
+    def _retrieve_latest_task_status(self):
+        """
+        Retrieve the latest task status.
+        """
+        if not self._task_exists():
+            raise ValueError(f"Task does not exist for {self.kbase_collection}, {self.load_ver}, {self.tool}")
+
+        latest_task = self._get_latest_task()
+        job_id = latest_task["job_id"]
+        job_status = self._get_job_status_from_nersc(job_id)
+
+        return job_status, job_id
+
+    def _check_preconditions(self):
+        """
+        Check conditions from previous runs of the same tool and load version
+        """
+
+        if not self._task_exists():
+            return True
+
+        latest_task = self._get_latest_task()
+        latest_task_status, job_id = self._retrieve_latest_task_status()
+
+        if self.force_run:
+            # cancel the previous job if it is running or pending
+            if latest_task_status in [JobStatus.RUNNING, JobStatus.PENDING]:
+                self._cancel_job(job_id)
+            return True
+
+        if latest_task['source_data_dir'] != self.source_data_dir:
+            raise PreconditionError(
+                f'There is a previous run of the same tool and load version with a different source data directory.')
+
+        if latest_task_status in [JobStatus.FAILED, JobStatus.CANCELLED]:
+            return True
+        elif latest_task_status in [JobStatus.COMPLETED, JobStatus.RUNNING, JobStatus.PENDING]:
+            raise PreconditionError(
+                f'There is a previous run of the same tool and load version that is {str(latest_task_status)}.')
+        else:
+            raise ValueError(f'Unexpected job status: {latest_task_status}')
+
+    def __init__(self, kbase_collection, load_ver, tool, source_data_dir, root_dir=loader_common_names.ROOT_DIR,
+                 force_run=False):
+        """
+        Initialize the task manager.
+
+        :param kbase_collection: KBase collection identifier name (e.g. GTDB).
+        :param load_ver: collection load version (e.g. r207.kbase.1).
+        :param tool: tool name (e.g. gtdb_tk, checkm2, etc.)
+        :param source_data_dir: source data directory.
+        :param root_dir: root directory of the collection project.
+                         Default is the ROOT_DIR defined in src/loaders/common/loader_common_names.py
+        :param force_run: if True, remove contents of the old job directory and run the job.
+        """
+
+        self.kbase_collection = kbase_collection
+        self.load_ver = load_ver
+        self.tool = tool
+        self.source_data_dir = source_data_dir
+        self.root_dir = root_dir
+        self.force_run = force_run
+
+        # job directory is named as <kbase_collection>_<load_ver>_<tool>
+        self.job_dir = os.path.join(self.root_dir, TASKFARMER_JOB_DIR,
+                                    f'{self.kbase_collection}_{self.load_ver}_{self.tool}')
+
+        self._create_job_dir(destroy_old_job_dir=self.force_run)
```

Comment:
This is another item where I'm sort of thinking ahead a bit, but theoretically there could be some time between creation of the manager and a submit_job call. That means there could be a case where there is a currently running job and the manager is created and deletes the job directory, which could cause lots of weird failures.

Maybe the force_run argument should be provided to submit_job and directory deleted after check preconditions has run, so the prior job is cancelled first

Reply:
👍

Reply:
I found a bug after making submit_job take force_run and recreate job_dir at the check precondition step. We actually want to recreate the folder before all scripts are generated which is done by task_generator.py.

Reply:
Thinking about this a bit more, the task manager isn't responsible for setting up the job dir - that's the responsibility of the calling script. Maybe the task manager shouldn't delete the job dir at all and force (or maybe a better named argument) just kills any currently running job

Reply:
👍

Reply:
:+1: